Book List

lmaroso opened the pull request:
# Book List





## Summary

Added a Book List to my Todo project using React Navigation.





## Book List functionalities

- _Books are loaded from an [API](https://private-69a0f-lucasmaroso.apiary-mock.com/books)_
- _If the book has no image, it will load a placeholder.._





## Screenshots

#### Book List screen

![Screenshot_20190807-091052](https://user-images.githubusercontent.com/51789658/62621943-83d74e00-b8f3-11e9-912a-f32528694d40.png)

#### Spinner

![Screenshot_20190807-091054](https://user-images.githubusercontent.com/51789658/62621955-8b96f280-b8f3-11e9-970e-178e40c1d07c.png)

Commit:
Adding some clean up on old functions.
Created a list of books.

Commit:
Adding Async functions to recover API's data.
I push the code to continue working at home.

Commit:
Listing books on a Scrollview.
It's not done with redux yet.

Commit:
Apply fixes of previous commit
 - Fix Santi's requirements of previous PR.
 - Usage of bold font on book's titles.
 - Reduction of book's images.
 - Addition of books to redux store (not working)

Commit:
Added bookList to redux store.
Added spinner while getting books from API.

Commit:
Corrected various bugs and added some improvements
Correction of PropTypes.
Handled API request errors using modals.
Increased spinner size.
Added reload button to send new request.
Added placeholder if book has no image.
unification of Book's and Todo's Lists.
Fixed book title's size problem (exceeded View size).

Commit:
Removed redux-thunk.

sdandois commented on src/Router/routes.js:
```diff
@@ -6,8 +6,9 @@ const generateRoutes = routeNamesArray => {
+const STACK_ROUTE_LIST = ['Home'];
 
-const Routes = generateRoutes(ROUTE_LIST);
+const Routes = generateRoutes(TAB_ROUTE_LIST, STACK_ROUTE_LIST);
```
I am not sure if this is working correctly. Check that generateRoutes takes one `parameter`  and you are giving it two. The second one is probably never read.

Check the contents of `Routes` `with console.log(Routes)`

lmaroso replied:
You're right.
Changed on next commit!

sdandois commented on src/components/Button/index.js:
```diff
@@ -4,16 +4,16 @@ import PropTypes from 'prop-types';
 };
 
-export default DeleteSelectedButton;
+export { Button };
```
Use:  `export default Button;`

lmaroso replied:
Done on the next commit.

sdandois commented on src/components/ErrorModal/index.js:
```diff
@@ -0,0 +1,30 @@
+import styles from './styles';
+
+const ErrorModal = ({ error }) => {
+  const [visible, setVisible] = React.useState(true);
```
Oh, oh, we have a hooks lover here. You should gather around with Roy.

It's good that you learn this tools, but we don't usually use it in the projects because they are new and appeared recently in the version `0.59`. The same functionality can  be achieved with a class and a state.

lmaroso replied:
Haha! Not really familiar with hooks yet.
Yesterday we saw this with Roy and thunk it was a good idea to use it.
I hope to learn this tool soon as it's really practical in situations like this (using flags, for example).

Commit:
Correction of an import that has an invalid route.

sdandois commented on src/components/List/List.js:
```diff
@@ -1,30 +1,29 @@
-    />
-  );
+const List = ({ elems, render }) => {
+  return <FlatList keyExtractor={elem => elem.id} data={elems} renderItem={render} />;
```
This Component is too simple, you should use the FlatList directly.

lmaroso replied:
Done on the next commit.

sdandois commented on src/screens/Books/components/ListItem/index.js:
```diff
@@ -0,0 +1,37 @@
+import { Text, View, Image } from 'react-native';
+import PropTypes from 'prop-types';
+
+import { CardSection } from '../../../../components/CardSection';
```
Use default export in components.

In `CardSection.js` file, do: `export default CardSection;`

And here you do: `import CardSection from ...`

lmaroso replied:
Done on the next commit.

sdandois commented on src/screens/Books/components/ListItem/styles.js:
```diff
@@ -0,0 +1,33 @@
+  headerContent: {
+    flex: 1,
+    flexDirection: 'column',
+    justifyContent: 'space-around'
```
Use `flex start` here and add some padding to the headerText

lmaroso replied:
Done on the next commit.

sdandois commented on src/screens/Books/index.js:
```diff
@@ -0,0 +1,82 @@
+
+import { actionCreators as bookActions } from '../../redux/books/actions';
+import { Spinner } from '../../components/Spinner';
+import { ErrorModal } from '../../components/ErrorModal';
```
Use default export in components `Spinner` and `ErrorModal`.

lmaroso replied:
Done on the next commit.

sdandois commented on src/screens/Books/index.js:
```diff
@@ -0,0 +1,82 @@
+
+class Books extends Component {
+  async componentDidMount() {
+    this.loadContent();
```
I think you should either return `this.loadContet()` or either await for it.
If you don't, `this.loadContent` generates a promise that you are not waiting for it to finish.

lmaroso replied:
To solve this, I removed async from `componentDidMount()` as `loadContent()` is an async function.

sdandois commented on src/screens/Books/index.js:
```diff
@@ -0,0 +1,82 @@
+        </View>
+      );
+    }
+    if (error !== null) {
```
```suggestion
    else if (error !== null) {
```

lmaroso replied:
I removed the "loading" if statement on the next commit.

sdandois commented on src/screens/Books/index.js:
```diff
@@ -0,0 +1,82 @@
+    }
+    return (
+      <View style={styles.textButtonView}>
+        <List elems={books} render={({ item }) => <ListItem elem={item} />} />
```
This component `List` has no existene purpose. Remove it and use a flat list here.

You can make a class method named `renderItem` that does what here render does.

lmaroso replied:
Done on the next commit.

sdandois commented on src/screens/Books/index.js:
```diff
@@ -0,0 +1,82 @@
+    return (
+      <View style={styles.textButtonView}>
+        <List elems={books} render={({ item }) => <ListItem elem={item} />} />
+        <TextButton onPress={() => this.loadContent()} title="Refresh" style={styles.textButton} />
```
Try to avoid creating arrow functions during the render method. Here you can pass the function directly if you bind this correctly.

lmaroso replied:
Okay! This if statement is removed on the next commit.

sdandois reviewed (changes requested):
Great Work! Check my comments.

Also, add a text and a button in case of error.

Commit:
Changed some things requested by Santi:
 - Bug Fixed:
     - Imported ListItem from Todo screen has an invalid route.
     - generateRoutes function received 2 parameters and accepts one.
 - Changes:
     - Removed the refresh button from BookList. Now, the refresh function is handled on a flatlist.
     - CardSection, ErrorModal and Button are exported as default.
     - Added paddingBottom to Title.
     - Deleted List component. Now, flatlist is implemented on the screens.
     - Add paddingHorizontal on ErrorModal.
     - If an error occurred, it shows a modal with an "accept" button. After pressing it, you can see a text with an error message and a button to reload.

lmaroso commented:
#### View changes on BookList

![Screenshot_1565203366](https://user-images.githubusercontent.com/51789658/62649592-7c339b80-b92b-11e9-9379-1fffe47dedde.png)


#### Spinner managed by FlatList

![Screenshot_1565203364](https://user-images.githubusercontent.com/51789658/62649612-88b7f400-b92b-11e9-9b1f-829d2f77844e.png)


#### Now, if you scroll down the book list, it will refresh the books.

![Screenshot_1565203378](https://user-images.githubusercontent.com/51789658/62649626-91a8c580-b92b-11e9-9c3e-12a656f6131a.png)

#### If an error occurred while getting the books, it will show a modal with the error message. 

![Screenshot_1565203814](https://user-images.githubusercontent.com/51789658/62649669-a1c0a500-b92b-11e9-9452-8460b39ac501.png)

#### After pressing the `accept` button, it will show a screen with a text saying that an error occurred and a refresh button.

![Screenshot_1565203817](https://user-images.githubusercontent.com/51789658/62649691-aab17680-b92b-11e9-8037-3b26fc7c01dd.png)

sdandois commented on src/screens/Todo/index.js:
```diff
@@ -0,0 +1,55 @@
+import ListItem from './components/ListItem';
+import styles from './styles';
+
+// const Todo = ({ todos, renderItem, dispatchRemoveSelected }) => {
```
Remove this comment when possible.

sdandois reviewed (approved):
Great! :smile:
